# Task 1

## In this task I made a solution only for the batche **JetBrainsEUR_settlement_detail_report_batch_138**, all other batches with a difference in the table will be solved in the same way, just with different csv files

In [1]:
import pandas as pd


In [151]:
csv_file = 'JetBrainsEUR_settlement_detail_report_batch_138.csv'
df = pd.read_csv(csv_file, sep = ';')


In [152]:
df.head()

COMPANY_ACCOUNT MERCHANT_ACCOUNT  BATCH_NUMBER     PSP_REFERENCE  \
0       JetBrains     JetBrainsEUR           138  9E6F1E95657D4486   
1       JetBrains     JetBrainsEUR           138  D4752B4259E44AF9   
2       JetBrains     JetBrainsEUR           138  2C4AEF8756DC4086   
3       JetBrains     JetBrainsEUR           138  E70F68131302432A   
4       JetBrains     JetBrainsEUR           138  871E14EA5D0C4375   

    ORDER_REF PAYMENT_METHOD        DATE TIMEZONE     TYPE  \
0  A000051401           visa  2018-03-23      UTC  Settled   
1  E000088071           visa  2018-04-10      UTC  Settled   
2  E000029631             mc  2018-04-14      UTC  Settled   
3  R000006739           visa  2018-01-06      UTC  Settled   
4  A000004154             mc  2018-01-19      UTC  Settled   

  MODIFICATION_REFERENCE CURRENCY    NET   FEE  GROSS  
0       34FAE3CA4E924B2A      EUR  152.0   8.0  160.0  
1       BE7053CF74B2414B      EUR  344.0  18.0  362.0  
2       2ACA3E10AE39490D      EUR  317.0  17.0  334.0  
3       48E055D7A6D6470D      EUR  486.0  26.0  512.0  
4       01439DDE302945BB      EUR  475.0  25.0  500.0

In [153]:
total_net = df['GROSS'].sum()
print(f"Total Net Amount: {total_net}")


Total Net Amount: 11008626.0


## SQL code
### for getting **netsuite data** (doubleclick to view the code in normal way)

SELECT
    t.TRANSACTION_ID,
    t.TRANID,
    t.TRANSACTION_TYPE,
    t.TRANDATE,
    t.EXCHANGE_RATE,
    t.CURRENCY_ID,
    t.ORDER_REF,
    t.MERCHANT_ACCOUNT,
    t.BATCH_NUMBER,
    t.IS_NON_POSTING,
    t.DATE_LAST_MODIFIED,
    tl.TRANSACTION_LINE_ID,
    tl.AMOUNT,
    tl.AMOUNT_FOREIGN,
    tl.SUBSIDIARY_ID,
    tl.ACCOUNT_ID,
    tl.COMPANY_ID,
    tl.DATE_LAST_MODIFIED
FROM
    netsuite.TRANSACTIONS AS t
JOIN
    netsuite.TRANSACTION_LINES AS tl
ON
    t.TRANSACTION_ID = tl.TRANSACTION_ID
WHERE
    t.BATCH_NUMBER = 138
    and t.MERCHANT_ACCOUNT = 'JetBrainsEUR'

In [154]:
netsuite_data = pd.read_csv('netsuite_138.csv')

In [155]:
netsuite_data

TRANSACTION_ID      TRANID  TRANSACTION_TYPE    TRANDATE  \
0               741593  CP00079677           Payment  2018-04-13   
1               741593  CP00079677           Payment  2018-04-13   
2               741647  CP00074790           Payment  2018-04-07   
3               741647  CP00074790           Payment  2018-04-07   
4               741656  CP00013127           Payment  2018-01-20   
...                ...         ...               ...         ...   
131797         1484497  CD00000068  Customer Deposit  2018-03-14   
131798         1484511  CD00000098  Customer Deposit  2018-04-14   
131799         1484511  CD00000098  Customer Deposit  2018-04-14   
131800         1484537  CD00000076  Customer Deposit  2018-03-20   
131801         1484537  CD00000076  Customer Deposit  2018-03-20   

        EXCHANGE_RATE  CURRENCY_ID   ORDER_REF MERCHANT_ACCOUNT  BATCH_NUMBER  \
0            0.808244            3  Z000160678     JetBrainsEUR           138   
1            0.808244            3  Z000160678     JetBrainsEUR           138   
2            0.816427            3  E000047155     JetBrainsEUR           138   
3            0.816427            3  E000047155     JetBrainsEUR           138   
4            0.817294            3  Z000164632     JetBrainsEUR           138   
...               ...          ...         ...              ...           ...   
131797       0.810340            3  E000072223     JetBrainsEUR           138   
131798       0.811063            3  R000023898     JetBrainsEUR           138   
131799       0.811063            3  R000023898     JetBrainsEUR           138   
131800       0.813240            3  C000054418     JetBrainsEUR           138   
131801       0.813240            3  C000054418     JetBrainsEUR           138   

       IS_NON_POSTING DATE_LAST_MODIFIED  TRANSACTION_LINE_ID  AMOUNT  \
0                  No         2018-05-05                    0  -42.84   
1                  No         2018-05-05                    1   42.84   
2                  No         2018-04-12                    0  -62.05   
3                  No         2018-04-12                    1   62.05   
4                  No         2018-02-05                    0  -46.59   
...               ...                ...                  ...     ...   
131797             No         2018-04-11                    1   -1.62   
131798             No         2018-05-11                    0   25.95   
131799             No         2018-05-11                    1  -25.95   
131800             No         2018-04-03                    0    1.63   
131801             No         2018-04-03                    1   -1.63   

        AMOUNT_FOREIGN  SUBSIDIARY_ID  ACCOUNT_ID  COMPANY_ID  \
0                -53.0              1         124        8190   
1                 53.0              1        2198        8190   
2                -76.0              1         124        5783   
3                 76.0              1        2198        5783   
4                -57.0              1         124       10234   
...                ...            ...         ...         ...   
131797            -2.0              1        2274        9158   
131798            32.0              1         124        8041   
131799           -32.0              1        2274        8041   
131800             2.0              1         124        5983   
131801            -2.0              1        2274        5983   

       DATE_LAST_MODIFIED.1  
0                2018-05-05  
1                2018-05-05  
2                2018-04-12  
3                2018-04-12  
4                2018-02-05  
...                     ...  
131797           2018-04-11  
131798           2018-05-11  
131799           2018-05-11  
131800           2018-04-03  
131801           2018-04-03  

[131802 rows x 18 columns]

In [156]:
merged_data = df.merge(netsuite_data, left_on=['ORDER_REF'], right_on=['ORDER_REF'], suffixes=('_csv', '_netsuite'))


In [157]:
merged_data

COMPANY_ACCOUNT MERCHANT_ACCOUNT_csv  BATCH_NUMBER_csv  \
0            JetBrains         JetBrainsEUR               138   
1            JetBrains         JetBrainsEUR               138   
2            JetBrains         JetBrainsEUR               138   
3            JetBrains         JetBrainsEUR               138   
4            JetBrains         JetBrainsEUR               138   
...                ...                  ...               ...   
132149       JetBrains         JetBrainsEUR               138   
132150       JetBrains         JetBrainsEUR               138   
132151       JetBrains         JetBrainsEUR               138   
132152       JetBrains         JetBrainsEUR               138   
132153       JetBrains         JetBrainsEUR               138   

           PSP_REFERENCE   ORDER_REF PAYMENT_METHOD        DATE TIMEZONE  \
0       9E6F1E95657D4486  A000051401           visa  2018-03-23      UTC   
1       9E6F1E95657D4486  A000051401           visa  2018-03-23      UTC   
2       9E6F1E95657D4486  A000051401           visa  2018-03-23      UTC   
3       9E6F1E95657D4486  A000051401           visa  2018-03-23      UTC   
4       D4752B4259E44AF9  E000088071           visa  2018-04-10      UTC   
...                  ...         ...            ...         ...      ...   
132149  8776791E5DFA4A0C  R000051461           visa  2018-01-20      UTC   
132150  981BD5A2B775465F  R000115379           visa  2018-01-22      UTC   
132151  981BD5A2B775465F  R000115379           visa  2018-01-22      UTC   
132152  981BD5A2B775465F  R000115379           visa  2018-01-22      UTC   
132153  981BD5A2B775465F  R000115379           visa  2018-01-22      UTC   

           TYPE MODIFICATION_REFERENCE  ... BATCH_NUMBER_netsuite  \
0       Settled       34FAE3CA4E924B2A  ...                   138   
1       Settled       34FAE3CA4E924B2A  ...                   138   
2       Settled       34FAE3CA4E924B2A  ...                   138   
3       Settled       34FAE3CA4E924B2A  ...                   138   
4       Settled       BE7053CF74B2414B  ...                   138   
...         ...                    ...  ...                   ...   
132149  Settled       F35A6970B54E4DFB  ...                   138   
132150  Settled       17C17DC50EFD4607  ...                   138   
132151  Settled       17C17DC50EFD4607  ...                   138   
132152  Settled       17C17DC50EFD4607  ...                   138   
132153  Settled       17C17DC50EFD4607  ...                   138   

        IS_NON_POSTING  DATE_LAST_MODIFIED  TRANSACTION_LINE_ID  AMOUNT  \
0                   No          2018-04-14                    0 -123.08   
1                   No          2018-04-14                    1  123.08   
2                   No          2018-04-01                    0   -6.48   
3                   No          2018-04-01                    1    6.48   
4                   No          2018-05-09                    0 -279.71   
...                ...                 ...                  ...     ...   
132149              No          2018-02-11                    1   13.08   
132150              No          2018-02-05                    0  -62.08   
132151              No          2018-02-05                    1   62.08   
132152              No          2018-01-30                    0   -3.27   
132153              No          2018-01-30                    1    3.27   

       AMOUNT_FOREIGN SUBSIDIARY_ID ACCOUNT_ID  COMPANY_ID  \
0              -152.0             1        124        1237   
1               152.0             1       2198        1237   
2                -8.0             1        124        1237   
3                 8.0             1       2274        1237   
4              -344.0             1        124        1235   
...               ...           ...        ...         ...   
132149           16.0             1       2274        2727   
132150          -76.0             1        124        2727   
132151           76.0    

In [158]:
filtered_data = merged_data[(merged_data['AMOUNT_FOREIGN'] > 0) & (merged_data['ACCOUNT_ID'] == 2198)]

In [159]:
filtered_data['GROSS_difference'] = filtered_data['GROSS'] - filtered_data['AMOUNT_FOREIGN']

<ipython-input-159-5b3828b23cc8>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_data['GROSS_difference'] = filtered_data['GROSS'] - filtered_data['AMOUNT_FOREIGN']


In [160]:
filtered_data.head()

COMPANY_ACCOUNT MERCHANT_ACCOUNT_csv  BATCH_NUMBER_csv     PSP_REFERENCE  \
1        JetBrains         JetBrainsEUR               138  9E6F1E95657D4486   
5        JetBrains         JetBrainsEUR               138  D4752B4259E44AF9   
9        JetBrains         JetBrainsEUR               138  2C4AEF8756DC4086   
13       JetBrains         JetBrainsEUR               138  E70F68131302432A   
17       JetBrains         JetBrainsEUR               138  871E14EA5D0C4375   

     ORDER_REF PAYMENT_METHOD        DATE TIMEZONE     TYPE  \
1   A000051401           visa  2018-03-23      UTC  Settled   
5   E000088071           visa  2018-04-10      UTC  Settled   
9   E000029631             mc  2018-04-14      UTC  Settled   
13  R000006739           visa  2018-01-06      UTC  Settled   
17  A000004154             mc  2018-01-19      UTC  Settled   

   MODIFICATION_REFERENCE  ... IS_NON_POSTING  DATE_LAST_MODIFIED  \
1        34FAE3CA4E924B2A  ...             No          2018-04-14   
5        BE7053CF74B2414B  ...             No          2018-05-09   
9        2ACA3E10AE39490D  ...             No          2018-04-30   
13       48E055D7A6D6470D  ...             No          2018-01-23   
17       01439DDE302945BB  ...             No          2018-01-26   

    TRANSACTION_LINE_ID  AMOUNT  AMOUNT_FOREIGN SUBSIDIARY_ID ACCOUNT_ID  \
1                     1  123.08           152.0             1       2198   
5                     1  279.71           344.0             1       2198   
9                     1  257.11           317.0             1       2198   
13                    1  402.73           486.0             1       2198   
17                    1  389.84           475.0             1       2198   

   COMPANY_ID  DATE_LAST_MODIFIED.1  GROSS_difference  
1        1237            2018-04-14               8.0  
5        1235            2018-05-09              18.0  
9        1236            2018-04-30              17.0  
13       1236            2018-01-23              26.0  
17       1235            2018-01-26              25.0  

[5 rows x 32 columns]

In [161]:
discrepancies = filtered_data[(filtered_data['GROSS_difference'] != 0)] 

In [162]:
discrepancies.head()

COMPANY_ACCOUNT MERCHANT_ACCOUNT_csv  BATCH_NUMBER_csv     PSP_REFERENCE  \
1        JetBrains         JetBrainsEUR               138  9E6F1E95657D4486   
5        JetBrains         JetBrainsEUR               138  D4752B4259E44AF9   
9        JetBrains         JetBrainsEUR               138  2C4AEF8756DC4086   
13       JetBrains         JetBrainsEUR               138  E70F68131302432A   
17       JetBrains         JetBrainsEUR               138  871E14EA5D0C4375   

     ORDER_REF PAYMENT_METHOD        DATE TIMEZONE     TYPE  \
1   A000051401           visa  2018-03-23      UTC  Settled   
5   E000088071           visa  2018-04-10      UTC  Settled   
9   E000029631             mc  2018-04-14      UTC  Settled   
13  R000006739           visa  2018-01-06      UTC  Settled   
17  A000004154             mc  2018-01-19      UTC  Settled   

   MODIFICATION_REFERENCE  ... IS_NON_POSTING  DATE_LAST_MODIFIED  \
1        34FAE3CA4E924B2A  ...             No          2018-04-14   
5        BE7053CF74B2414B  ...             No          2018-05-09   
9        2ACA3E10AE39490D  ...             No          2018-04-30   
13       48E055D7A6D6470D  ...             No          2018-01-23   
17       01439DDE302945BB  ...             No          2018-01-26   

    TRANSACTION_LINE_ID  AMOUNT  AMOUNT_FOREIGN SUBSIDIARY_ID ACCOUNT_ID  \
1                     1  123.08           152.0             1       2198   
5                     1  279.71           344.0             1       2198   
9                     1  257.11           317.0             1       2198   
13                    1  402.73           486.0             1       2198   
17                    1  389.84           475.0             1       2198   

   COMPANY_ID  DATE_LAST_MODIFIED.1  GROSS_difference  
1        1237            2018-04-14               8.0  
5        1235            2018-05-09              18.0  
9        1236            2018-04-30              17.0  
13       1236            2018-01-23              26.0  
17       1235            2018-01-26              25.0  

[5 rows x 32 columns]

In [165]:
total_diff = discrepancies['GROSS_difference'].sum()
print(f"Total Diff Amount: {total_diff}")

Total Diff Amount: 512598.0


## The sum of the differences in transactions in my code and in the table in the task of course do not match for this particular batches. But I assume that approximately in the same way it is possible to solve this problem when more business context and data is provided